src/java.desktop/macosx/classes/sun/lwawt/macosx/CAccessibility.java

```diff
@@ -757,14 +759,75 @@ private static Object[] getChildrenAndRolesImpl(Accessible a, Component c, int w
         return new Object[]{childrenAndRoles.get(whichChildren * 2), childrenAndRoles.get((whichChildren * 2) + 1)};
     }
 
+    private static Accessible createAccessibleTreeNode(JTree t, TreePath p) {
+        Accessible a = null;
+
+        try {
+            Class<?> accessibleJTreeNodeClass = Class.forName("javax.swing.JTree$AccessibleJTree$AccessibleJTreeNode");
+            Constructor<?> constructor = accessibleJTreeNodeClass.getConstructor(t.getAccessibleContext().getClass(), JTree.class, TreePath.class, Accessible.class);
+            constructor.setAccessible(true);
+            a = ((Accessible) constructor.newInstance(t.getAccessibleContext(), t, p, null));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return a;
+    }
+
     // This method is called from the native
     // Each child takes up three entries in the array: one for itself, one for its role, and one for the recursion level
     private static Object[] getChildrenAndRolesRecursive(final Accessible a, final Component c, final int whichChildren, final boolean allowIgnored, final int level) {
         if (a == null) return null;
         return invokeAndWait(new Callable<Object[]>() {
             public Object[] call() throws Exception {
-                ArrayList<Object> currentLevelChildren = new ArrayList<Object>();
                 ArrayList<Object> allChildren = new ArrayList<Object>();
+
+                Accessible at = null;
+                if (a instanceof CAccessible) {
+                    at = CAccessible.getSwingAccessible(a);
+                } else {
+                    at = a;
+                }
+
+                if (at instanceof JTree) {
+                    JTree tree = ((JTree) at);
+
+                    if (whichChildren == JAVA_AX_ALL_CHILDREN) {
+                        int count = tree.getRowCount();
+                        for (int i = 0; i < count; i++) {
+                            TreePath path = tree.getPathForRow(i);
+                            Accessible an = createAccessibleTreeNode(tree, path);
```

Comment:
I see that you are creating the accessible tree node with reflection calling its constructor. Can you just elaborate if it is the only way to get hold of it and why getting a tree.getAccessible().getAccessibleChild(i) does not work or what kind of problem getting the tree node this way creates?

Reply:
A little lower, the access algorithm is implemented in the way you showed.
But to speed things up, I want to avoid recursion through the native and for this it is easier to initialize those nodes that are present in the tree path through the constructor. Why do I propose to use reflection? Otherwise, you will either have to make the constructor of a double-nested class public, or create a chain of methods that allow you to instantiate this class through a new call in CAccessibility.

Reply:
I can't say i like the approach but i can accept it as a hotfix to eliminate the situation but only if we create a followup bug to fix it properly - even if it will require a new public accessor to be added to the JTree class that will instantiate the tree node accessibility peer for a given TreePath without the need to use a reflection. Using reflection like this really breaks the encapsulation and logic separation principles introducing the unneeded dependency on the implementation details in the unrelated class. So as a bottom line - i will approve that as a hotfix but the followup bug to properly eliminate the implementation dependency and getting rid of the reflection call to be filled and addressed in the next release. If that will require a CSR - let there be CSR.